### PR TITLE
Document default encoding/decoding behavior for Go to BSON

### PR DESCRIPTION
**Pull request by divjotarora**
A second attempt at some docs prework based on the meeting we had last week. The idea here is to enumerate the mappings from BSON to Go types for decoding into a D or M and list some information about struct behavior.

### CODE REVIEW

File: bson/doc.go [hunk 1]
```diff
@@ -19,24 +20,90 @@
...
+//       otherwise.
+//       4. int64 marshals to BSON int64.
+//       5. uint8 and uint16 marshal to a BSON int32.
+//       6. uint, uint32, and uint64 marshal to a BSON int32 if the value is between math.MinInt32 and math.MaxInt32,
+//       inclusive, and BSON int64 otherwise.
+//
+// Structs
+//
+// Structs can be marshalled/unmarshalled to/from BSON. When transforming structs to/from BSON, the following rules
+// apply:
```
Comment (iwysiu) [inline bson/doc.go]: This should mention that an interface{} will be unmarshalled as a bson.D

File: bson/doc.go [hunk 2]
```diff
@@ -19,24 +20,90 @@
...
+//
+//     3. An embedded struct field is marshalled as a subdocument. The key will be the lowercased name of the field's type.
+//
+//     4. A pointer field is marshalled as the underlying type if the pointer is non-nil. If the pointer is nil, it is
+//     marshalled as a BSON null value.
+//
+// The following struct tags can be used to configure behavior:
+//
+//     1. omitempty: If the omitempty struct tag is specified on a field, the field will not be marshalled if it is set to
+//     the zero value. By default, omitempty is not relevant for struct fields, which will be serialized as an embedded
```
Comment (iwysiu) [inline bson/doc.go]: In default mode omitempty still applies for structs that implement IsZero()

File: bson/doc.go [hunk 3]
```diff
@@ -19,24 +20,90 @@
...
+//     2. minsize: If the minsize struct tag is specified on a field of type int64, uint, uint32, or uint64 and the value of
+//     the field can fit in a signed int32, the field will be serialized as a BSON int32 rather than a BSON int64. For other
+//     types, this tag is ignored.
+//
+//     3. truncate: If the truncate struct tag is specified on a field with a non-float64 numeric type, BSON doubles unmarshalled
+//     into that field will be trucated at the decimal point. For example, if 3.14 is unmarshalled into a field of type int,
+//     it will be unmarshalled as 3. If this tag is not specified, the decoder will throw an error if the value cannot be
+//     decoded without losing precision. For float64 or non-numeric types, this tag is ignored.
+//
+//     4. inline: If the inline struct tag is specified for a struct or map field, the field will be "flattened" when
```
Comment (iwysiu) [inline bson/doc.go]: This should mention that only one inline map can be used at a time and that there can't be duplicate keys

File: bson/doc.go (hunk 1 above)
Comment (divjotarora) [reply bson/doc.go]: done.

File: bson/doc.go (hunk 2 above)
Comment (divjotarora) [reply bson/doc.go]: done.

File: bson/doc.go (hunk 3 above)
Comment (divjotarora) [reply bson/doc.go]: done.

File: bson/bsoncodec/time_codec.go [hunk 4]
```diff
@@ -22,7 +22,7 @@ const (
 
 var defaultTimeCodec = NewTimeCodec()
 
-// TimeCodec is the Codec used for struct values.
+// TimeCodec is the Codec used for time.Time values.
```
Comment (jyemin) [inline bson/bsoncodec/time_codec.go]: Whoops!

File: bson/doc.go [hunk 5]
```diff
@@ -4,8 +4,9 @@
 // not use this file except in compliance with the License. You may obtain
 // a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 
-// Package bson is a library for reading, writing, and manipulating BSON. The
-// library has two families of types for representing BSON.
+// Package bson is a library for reading, writing, and manipulating BSON.
```
Comment (jyemin) [inline bson/doc.go]: Perhaps link to bsonspec.org or https://docs.mongodb.com/manual/reference/bson-types/?

Comment (jyemin) [inline bson/doc.go]: Consider a one sentence description of BSON, perhaps just copied from https://docs.mongodb.com/manual/reference/bson-types/

File: bson/doc.go [hunk 6]
```diff
@@ -19,24 +20,96 @@
 // 		i32, ok := val.Int32OK()
 // 		// do something with i32...
 //
-// The D family of types is used to build concise representations of BSON using native Go types.
-// These types do not support automatic lookup.
+// Native Go Types
```
Comment (jyemin) [inline bson/doc.go]: What makes D and M native, in comparison to bson.Raw?  Do you mean that they are aliases for Go slice and map, respectively?

File: bson/doc.go [hunk 7]
```diff
@@ -19,24 +20,96 @@
...
+//
+// This D and M types defined in this package can be used to build concise representations of BSON using native Go
```
Comment (jyemin) [inline bson/doc.go]: By concise, do you mean that the code is concise or the representation in memory is concise?  Likely the former, but it reads like you mean the latter.  Consider saying directly that M is a map and D is a slice.

File: bson/doc.go [hunk 8]
```diff
@@ -19,24 +20,96 @@
...
+// types.
 //
 // Example:
```
Comment (jyemin) [inline bson/doc.go]: Since this section is for D and M, add an example of M as well.

File: bson/doc.go (hunk 7 above)
Comment (jyemin) [inline bson/doc.go]: Consider a sentence describing when you definitely need to use D (i.e. for commands)

File: bson/doc.go [hunk 9]
```diff
@@ -19,24 +20,96 @@
...
+//
+//     5. When unmarshalling, a field of type interface{} will follow the D/M type mappings listed above. BSON documents
+//     unmarshalled into an interface{} field will be unmarshalled as a D.
+//
+// The following struct tags can be used to configure behavior:
+//
+//     1. omitempty: If the omitempty struct tag is specified on a field, the field will not be marshalled if it is set to
+//     the zero value. By default, a struct field is only considered empty if the field's type implements the Zeroer
+//     interface and the IsZero method returns true. If the field's type does not implement Zeroer, the field will be
+//     marshalled as an embedded document.
```
Comment (jyemin) [inline bson/doc.go]: Even for a scalar value?  Or does this only apply to non-scalar values (i.e. arrays and documents)?

File: bson/doc.go [hunk 10]
```diff
@@ -19,24 +20,96 @@
...
+//
+//     2. minsize: If the minsize struct tag is specified on a field of type int64, uint, uint32, or uint64 and the value of
+//     the field can fit in a signed int32, the field will be serialized as a BSON int32 rather than a BSON int64. For other
+//     types, this tag is ignored.
+//
+//     3. truncate: If the truncate struct tag is specified on a field with a non-float64 numeric type, BSON doubles unmarshalled
```
Comment (jyemin) [inline bson/doc.go]: What about float32?

File: bson/doc.go [hunk 11]
```diff
@@ -19,24 +20,96 @@
...
+//     pulled up one level and will become top-level fields rather than being fields in a nested document. For example, if a
+//     map field named "Map" with value map[string]interface{}{"foo": "bar"} is inlined, the resulting document will be
+//     {"foo": "bar"} instead of {"map": {"foo": "bar"}}. There can only be one inlined map field in a struct. If there are
+//     duplicated fields in the resulting document when an inlined field is marshalled, an error will be returned. This tag
+//     can be used with fields that are pointers to structs. If an inlined pointer field is nil, it will not be marshalled.
+//     For fields that are not maps or structs, this tag is ignored.
+//
+// Marshalling and Unmarshalling
+//
+// Manually marshallhing and unmarshalling can be done with the Marshal and Unmarshal family of functions. To read or
```
Comment (jyemin) [inline bson/doc.go]: marshallhing => marshalling

File: bson/doc.go [hunk 12]
```diff
@@ -19,24 +20,96 @@
...
+// write BSON from a non-slice source, an Encoder or Decoder with a bsonrw.ValueWriter or bsonrw.ValueReader must be
```
Comment (jyemin) [inline bson/doc.go]: I'm not sure about this last sentence.  Will it make sense to most people in this context?  Just looking in marshal.go, I can't find the functions to which this sentence applies.

File: bson/doc.go (hunk 5 above)
Comment (divjotarora) [reply bson/doc.go]: done.

Comment (divjotarora) [reply bson/doc.go]: done.

File: bson/doc.go (hunk 6 above)
Comment (divjotarora) [reply bson/doc.go]: Yeah, the idea was that they're Go representations of BSON, rather than raw BSON bytes.

File: bson/doc.go (hunk 7 above)
Comment (divjotarora) [reply bson/doc.go]: It's supposed to mean that the code is concise, but I can see how that's confusing. Changed to say that M is a map and D is a slice.

Comment (divjotarora) [reply bson/doc.go]: This is described where the types are defined and I'd prefer not to duplicate when possible. Added a note to say users should refer to the type definitions for use cases and more info.

File: bson/doc.go (hunk 8 above)
Comment (divjotarora) [reply bson/doc.go]: done.

File: bson/doc.go (hunk 9 above)
Comment (divjotarora) [reply bson/doc.go]: This was supposed to be a continuation of the previous sentence. Clarified this to explicitly say *structs* that do not implement `Zeroer` will be marshalled as embedded documents.

File: bson/doc.go (hunk 10 above)
Comment (divjotarora) [reply bson/doc.go]: This was supposed to be for float32 and float64, not just float64. Fixed to say "non-float numeric types"

File: bson/doc.go (hunk 11 above)
Comment (divjotarora) [reply bson/doc.go]: done.

File: bson/doc.go (hunk 12 above)
Comment (divjotarora) [reply bson/doc.go]: I took this from the previous docs, but I'd be surprised if this ever applies to anyone. I don't think it applies to any of those functions. To do this, you'd have to create an implementation of `ValueWriter` or `ValueReader` and then use those to create an `Encoder` or `Decoder`. Given all of those, I think it's best to remove these comments altogether.

File: bson/doc.go (hunk 6 above)
Comment (jyemin) [reply bson/doc.go]: Ack

File: bson/doc.go [hunk 13]
```diff
@@ -22,11 +23,13 @@
 //
 // Native Go Types
 //
-// This D and M types defined in this package can be used to build concise representations of BSON using native Go
-// types.
+// This D and M types defined in this package can be used to build concise of BSON using native Go types. D is a slice
```
Comment (jyemin) [inline bson/doc.go]: No longer grammatical.  Concise has to modify something.

File: bson/doc.go [hunk 14]
```diff
@@ -86,14 +89,14 @@
 //
 //     1. omitempty: If the omitempty struct tag is specified on a field, the field will not be marshalled if it is set to
 //     the zero value. By default, a struct field is only considered empty if the field's type implements the Zeroer
-//     interface and the IsZero method returns true. If the field's type does not implement Zeroer, the field will be
-//     marshalled as an embedded document.
+//     interface and the IsZero method returns true. Struct fields of types that do not implement Zeroer will be
+//     marshalled as embedded documents.
```
Comment (jyemin) [inline bson/doc.go]: Struct fields of types that do implement Zeroer are also marshalled as embedded document (when IsZero is false).  Do you mean something like:

    Struct fields of types that do not implement Zeroer are _always_ marshalled as embedded documents

?

File: bson/doc.go (hunk 13 above)
Comment (divjotarora) [reply bson/doc.go]: Whoops. Changed to "The D and M types defined in this package can be used to build representations of BSON using native Go types"

File: bson/doc.go (hunk 14 above)
Comment (divjotarora) [reply bson/doc.go]: done.

**Pull request merged**